### PR TITLE
Fixes + Support of using DU inside kernels

### LINKED ISSUE #114: Support for loop with step
**Is your feature request related to a problem? Please describe.**
```
Unhandled exception. System.Exception: Unsupported call: op_rangestep
   at Microsoft.FSharp.Core.PrintfModule.PrintFormatToStringThenFail@1645.Invoke(String message)
   at Brahma.FSharp.OpenCL.Translator.Body.Translate(FSharpExpr expr, TargetContext`2 targetContext)
   at Brahma.FSharp.OpenCL.Translator.Body.TranslateAsExpr(FSharpExpr expr, TargetContext`2 targetContext)
   at Brahma.FSharp.OpenCL.Translator.Body.translateCond(FSharpExpr cond, TargetContext`2 targetContext)
   at Brahma.FSharp.OpenCL.Translator.Body.translateBinding(FSharpVar var, String newName, FSharpExpr expr, TargetContext`2 targetContext)
   at Brahma.FSharp.OpenCL.Translator.Body.translateLet(FSharpVar var, FSharpExpr expr, FSharpExpr inExpr, TargetContext`2 targetContext)
   at Brahma.FSharp.OpenCL.Translator.Body.Translate(FSharpExpr expr, TargetContext`2 targetContext)
   at Brahma.FSharp.OpenCL.Translator.Body.translateLet(FSharpVar var, FSharpExpr expr, FSharpExpr inExpr, TargetContext`2 targetContext)
   at Brahma.FSharp.OpenCL.Translator.Body.Translate(FSharpExpr expr, TargetContext`2 targetContext)
   at <StartupCode$Brahma-FSharp-OpenCL-Translator>.$Translator.go$cont@153(IEnumerable`1 translatorOptions, TargetContext`2 context, FSharpExpr expr, FSharpList`1 vars, Unit unitVar)
   at <StartupCode$Brahma-FSharp-OpenCL-Translator>.$Translator.go@148-13(IEnumerable`1 translatorOptions, TargetContext`2 context, FSharpExpr expr, FSharpList`1 vars)
   at Brahma.FSharp.OpenCL.Translator.FSQuotationToOpenCLTranslator.translate(FSharpExpr qExpr, IEnumerable`1 translatorOptions)
   at <StartupCode$Brahma-FSharp-OpenCL-Translator>.$Translator.action@1(FSQuotationToOpenCLTranslator this, FSharpExpr qExpr, IEnumerable`1 translatorOptions, Unit _arg1)
   at Brahma.FSharp.OpenCL.Translator.FSQuotationToOpenCLTranslator.Translate(FSharpExpr qExpr, IEnumerable`1 translatorOptions)
   at Brahma.FSharp.OpenCL.Core.CLCodeGenerator.GenerateKernel(FSharpExpr lambda, ComputeProvider provider, ICLKernel kernel, IEnumerable`1 translatorOptions)
   at Brahma.FSharp.OpenCL.Core.ComputeProvider.CompileQuery[T](ComputeProvider this, FSharpExpr lambda, IEnumerable`1 translatorOptions, FSharpList`1 additionalSources)
   at Brahma.FSharp.OpenCL.Core.ComputeProvider.Compile[TRange,a](ComputeProvider this, FSharpExpr`1 query, FSharpOption`1 _options, FSharpOption`1 translatorOptions, FSharpOption`1 _outCode, FSharpOption`1 kernelName, FSharpOption`1 _additionalSources)
```
when
```fsharp
let kernel =
    <@
        fun (range: _1D) ->
            let mutable mut = range.GlobalID0
            for i in 0 .. 2 .. 10 do
                mut <- 0
    @>

opencl {
    do! RunCommand kernel (fun kernelPrepare -> kernelPrepare <| _1D(10))
}
|> OpenCLEvaluationContext().RunSync
```

**Additional context**
Version: 2.0.0-alpha6.1
Framework: netcoreapp3.1

### LINKED ISSUE #123: OpenCL DLL path is hardcoded, dll could not be loaded properly on Non-Debian Linux distros
**Describe the bug**
Brahma is looking for OpenCL DLLs at specific directory `/usr/lib/x86_64-linux-gnu` wich exists only on [Debian-based](https://unix.stackexchange.com/questions/43190/where-did-usr-lib64-go-and-what-is-usr-lib-x86-64-linux-gnu) linux distributions.

The specific path in Debian is used for [Multiarch](http://wiki.debian.org/Multiarch/) libraries.

The proper location of `libOpenCL.so.1.0.0` dll may vary in some distributions  (for some distros the default path is `/usr/lib/libOpenCL.so.1.0.0`), nevertheless currently it's not possible to specify exact location in any way, because it's hardcoded :point_right:  [here](https://github.com/YaccConstructor/Brahma.FSharp/blob/17d2f3e5a7d193f6b610e5f95399e8a833006759/src/YC.OpenCL.NET/Cl.API.cs#L46) 

So, execution fails with:
image

**To Reproduce**
Running Brahma.FSharp (checked versions: `alpha7.1` - `alpha9.1`) on non-debian Linux distributions will result in an error:

`
"Unable to load shared library '/usr/lib/x86_64-linux-gnu/libOpenCL.so.1.0.0' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: /usr/lib/x86_64-linux-gnu/libOpenCL.so.1.0.0: cannot open shared object file: No such file or directory"
`

**Could be hot-fixed** by linking dlls

```bash
sudo ln -s /usr/lib/libOpenCL.so.1.0.0 /usr/lib/x86_64-linux-gnu/libOpenCL.so.1.0.0
```
This "hot-fix" probably needs to be documented somewhere :confused: to help early-adopters

### LINKED ISSUE #124: Provide a way to specify exact path to OpenCL dll
**Description**
There are a huge number of different Linux distributions, custom Linux builds, and so on, so the location of `libOpenCL.so.1.0.0` may vary.

I think it will be convenient for a library user to be able to specify exact OpenCL DLL path through environment variable (e.g. `BrahmaOCLdllPath="usr/lib/libOpenCL.so.1.0.0"`) or explicitly by passing a path-string in code.

This could help a lot when using Brahma in _uncommon setup_ (not sure if there is such thing as _common setup_ in Linux)

**Related to**
[Related issue](https://github.com/YaccConstructor/Brahma.FSharp/issues/123)

**Alternatives considered**
_ _ _

Provide pre-configuration documentation for others Linux systems
e.g.

Here is my take on making Brahma.FSharp work with proper `libOpenCL.so.1.0.0`
_ _ _

The instructions bellow were tested on [EndevourOS](https://endeavouros.com/), which is a popular Arch-based linux distribution - so all further steps are *most likely* the same for others arch-based distros.

**Installing opencl dependencies on arch**

Installing vendor OpenCL implementation and OpenCL ICD Loader 
```bash
sudo pacman -Syu nvidia nvidia-utils
sudo pacman -S ocl-icd opencl-headers opencl-nvidia intel-compute-runtime # I have both intel graphics and nvidia gpu
```

*[OPTIONAL]* Installing [clinfo](https://github.com/Oblomov/clinfo) - which is     a simple cli-tool that enumerates all possible (known) properties of the OpenCL platform and devices available on the system.
```bash
sudo pacman -S clinfo  
```

Installing OpenCL ICD loaders
```bash
sudo pacman -S lib32-ocl-icd lib32-opencl-nvidia # nvidia loader required for NVIDIA CUDA
```
**Configuring OpenCL ICD loader (`libOpenCl.so`)**
To ensure the ICD loader from the `ocl-icd` package is used we should create a file in /etc/ld.so.conf.d which adds /usr/lib to the dynamic program loader's search directories
```bash
$ sudo touch /etc/ld.so.conf.d/00-usrlib.conf
$ sudo echo "/usr/lib" > /etc/ld.so.conf.d/00-usrlib.conf
$ cat /etc/ld.so.conf.d/00-usrlib.conf # now let's check
/usr/lib
```
for more details ICD read [ArchWiki](https://wiki.archlinux.org/title/GPGPU#OpenCL_ICD_loader_(libOpenCL.so))

Now, after **rebooting** the machine - check if DLLs paths are similar to:
```bash
$ locate libOpenCL.so.1.0.0
/usr/lib/libOpenCL.so.1.0.0
/usr/lib32/libOpenCL.so.1.0.0
```

**Linking DLLs (+ aliasing std. Debian MultiArch DLL path)**

Creating symbolic link:
```bash
sudo mkdir -p /usr/lib/x86_64-linux-gnu
sudo ln -s /usr/lib/usr/lib/libOpenCL.so.1.0.0 /usr/lib/x86_64-linux-gnu/libOpenCL.so.1.0.0
```

After all is done - double-check sym-links, should be something like this:
```bash
$ ls -la /usr/lib/ | grep opencl # vendors implementations
drwxr-xr-x   2 root root      4096 окт 19 16:38 intel-opencl
lrwxrwxrwx   1 root root        21 сен 20 22:47 libnvidia-opencl.so -> libnvidia-opencl.so.1
lrwxrwxrwx   1 root root        26 сен 20 22:47 libnvidia-opencl.so.1 -> libnvidia-opencl.so.470.74
-rwxr-xr-x   1 root root  19461936 сен 20 22:47 libnvidia-opencl.so.470.74
lrwxrwxrwx   1 root root        21 мая 29 18:03 libopencl-clang.so -> libopencl-clang.so.12
-rwxr-xr-x   1 root root    982232 мая 29 18:03 libopencl-clang.so.12
```

```bash
$ ls -la /usr/lib/ | grep libOpenCL # DLLs (OpenCL ICD) which is used by Brahma 
lrwxrwxrwx   1 root root        18 авг 19 14:32 libOpenCL.so -> libOpenCL.so.1.0.0
lrwxrwxrwx   1 root root        18 авг 19 14:32 libOpenCL.so.1 -> libOpenCL.so.1.0.0
-rwxr-xr-x   1 root root    183664 авг 19 14:32 libOpenCL.so.1.0.0
```

```bash
$ ls -la /usr/lib/x86_64-linux-gnu/ # created symbolic link to OpenCl ICD DLL
total 140
drwxr-xr-x   2 root root   4096 окт 20 14:10 .
drwxr-xr-x 183 root root 135168 окт 20 14:27 ..
lrwxrwxrwx   1 root root     27 окт 20 14:10 libOpenCL.so.1.0.0 -> /usr/lib/libOpenCL.so.1.0.0
```

### PR DESCRIPTION
**Pull request by dpanfilyonok**
## Proposed Changes

Added:
- Bitwise XOR, Bitwise Negation
- Support of non-integer `for` loops and loops with step (fix #114)
- Support of more types of barriers (`barrierLocal`, `barrierGlobal`, `barrierFull`)
- Support of structs transferring
- Support of using DU inside kernels
- Support of using atomic inc and dec for all suitable types
- Way to specify exact path to OpenCL.dll using `BRAHMA_OCL_PATH` env variable (fix #124, fix #123) (_should be tested_)

Also
- Improved data transferring to unmanaged memory. Now, in parallel
- Simplified types translation

## Types of changes

What types of changes does your code introduce to Brahma.FSharp?

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] Build and tests pass locally
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/Brahma.FSharp.OpenCL.Core/CustomMarshaler.fs [hunk 1]
```diff
@@ -129,7 +150,7 @@ type CustomMarshaler<'a>() =
         size, mem
 
     member this.WriteToUnmanaged(array: 'a[], ptr: IntPtr) =
-        for j = 0 to array.Length - 1 do
+        Parallel.For(0, array.Length, fun j ->
```
Comment (gsvgit) [inline src/Brahma.FSharp.OpenCL.Core/CustomMarshaler.fs:153]: Why not ```Array.Parallel.iteri```?

File: src/Brahma.FSharp.OpenCL.Translator/Body.fs [hunk 2]
```diff
@@ -428,7 +448,8 @@ module rec Body =
         do! State.modify (fun context -> context.VarDecls.Clear(); context)
 
         for expr in linearized do
-            do! State.modify (fun context -> context.VarDecls.Clear(); context)
+            // NOTE тут что то сломалось :(
```
Comment (gsvgit) [inline src/Brahma.FSharp.OpenCL.Translator/Body.fs:451]: Hm....

File: src/Brahma.FSharp.OpenCL.Translator/Body.fs [hunk 3]
```diff
@@ -551,32 +641,88 @@ module rec Body =
...
+            let! type' = Type.translate onType
+            let value =
+                Expr.Call(
+                    Utils.makeGenericMethodCall [onType] <@ LanguagePrimitives.GenericOne<int> @>,
+                    List.empty
+                ).EvaluateUntyped().ToString()
+
+            return Const(type', value) :> Node<_>
+
+        // TODO convert to active pattern?
```
Comment (gsvgit) [inline src/Brahma.FSharp.OpenCL.Translator/Body.fs:654]: I guess, yes, it should be converted to active patterns.

File: src/Brahma.FSharp.OpenCL.Translator/QuotationTransformers/TransformMinMax.fs [hunk 4]
```diff
@@ -0,0 +1,42 @@
+namespace Brahma.FSharp.OpenCL.Translator.QuotationTransformers
```
Comment (gsvgit) [inline src/Brahma.FSharp.OpenCL.Translator/QuotationTransformers/TransformMinMax.fs:1]: Why so complex? Why we can not use direct mapping to OpenCL ```min``` and ```max```?

**Review by gsvgit — COMMENTED** [review]
Cool!

File: src/Brahma.FSharp.OpenCL.Translator/QuotationTransformers/TransformMinMax.fs (hunk 4 above)
Comment (dpanfilyonok) [reply src/Brahma.FSharp.OpenCL.Translator/QuotationTransformers/TransformMinMax.fs:1]: #130

**Comment by gsvgit** [comment]
@dpanfilyonok Please, check the tests: https://github.com/YaccConstructor/Brahma.FSharp/runs/4689985283?check_suite_focus=true Looks like problems with dependencies or reflection. Locally tests failed with the same problem. 
```
Error Message:
   The type initializer for '<StartupCode$Brahma-FSharp-Tests>.$Common' threw an exception.
  Stack Trace:
     at Brahma.FSharp.Tests.Common.get_context()
   at Brahma.FSharp.Tests.Common.checkResult[a](FSharpExpr`1 command, a[] inArr, a[] expectedArr) in /home/gsv/Projects/Brahma.FSharp/tests/Brahma.FSharp.Tests/Common.fs:line 24
   at CompositeTypesTests.structTests@193-4.Invoke(Unit _arg4) in /home/gsv/Projects/Brahma.FSharp/tests/Brahma.FSharp.Tests/CompositeTypesTests.fs:line 203
   at Expecto.Impl.execTestAsync@566-1.Invoke(Unit unitVar)
   at Microsoft.FSharp.Control.AsyncPrimitives.CallThenInvoke[T,TResult](AsyncActivation`1 ctxt, TResult result1, FSharpFunc`2 part2) in D:\workspace\_work\1\s\src\fsharp\FSharp.Core\async.fs:line 386
   at <StartupCode$FSharp-Core>.$Async.StartChild@1663-5.Invoke(AsyncActivation`1 ctxt) in D:\workspace\_work\1\s\src\fsharp\FSharp.Core\async.fs:line 1663
   at Microsoft.FSharp.Control.Trampoline.Execute(FSharpFunc`2 firstAction) in D:\workspace\_work\1\s\src\fsharp\FSharp.Core\async.fs:line 104
```

**Pull request merged**